### PR TITLE
Improve and simplify fits Section indexing.

### PR DESCRIPTION
**Pull request by anntzer**
Add support for indexing fits Sections with slices with nontrivial (step != 1) slices, with integer arrays, and with boolean arrays.

### CODE REVIEW

**Comment by anntzer** [comment]
I finally asked numpy-discussion and got the explanation for the ellipsis support.  It is now implemented.

For proper testing, I have to use `np.array_equal` instead of `(x == y).all()`.

Small backwards compatibility break: `data.section[indices]` now returns a scalar rather than a size-one array if there are as many indices as dimensions.

**Comment by embray** [comment]
I'm not worried about the backwards compat issue--the change there feels like the right thing to do in the first place.

File: astropy/io/fits/hdu/image.py [hunk 1]
```diff
@@ -734,109 +733,80 @@ def __init__(self, hdu):
...
         if not isinstance(key, tuple):
             key = (key,)
         naxis = len(self.hdu.shape)
-        if naxis < len(key):
-            raise IndexError('too many indices')
-        elif naxis > len(key):
-            key = key + (slice(None),) * (naxis - len(key))
+        return_scalar = (all(isinstance(k, (int, np.integer)) for k in key)
+                         and len(key) == naxis)
+        if not any(k is ... for k in key):
```
Comment (embray) [inline astropy/io/fits/hdu/image.py]: This is invalid syntax on Python 2 (and below as well).  You need to use `Ellipsis`--the `...` literal does not work outside of slice notation.

**Comment by embray** [comment]
Fix the Python 2 issue and mention these improvements in the changelog (I think under "new features" is fine) and this will be good to merge.

**Comment by embray** [comment]
Looks ready to go.  Thank you for all your work on these improvements!  I think this code will come in handy in the future as well (for aforementioned improvements to FITS data access).

**Pull request merged**